sys/drv/plic.c

```diff
@@ -0,0 +1,190 @@
+  (PLIC_CONTEXT_BASE + PLIC_CTXNUM_SV * PLIC_CONTEXT_STRIDE)
+
+#define PLIC_PRIORITY(n) (PLIC_PRIORITY_BASE + (n) * sizeof(uint32_t))
+#define PLIC_ENABLE_SV(n) (PLIC_ENABLE_BASE_SV + ((n) / 32) * sizeof(uint32_t))
+#define PLIC_THRESHOLD_SV (PLIC_CONTEXT_BASE_SV + PLIC_CONTEXT_THRESHOLD)
+#define PLIC_CLAIM_SV (PLIC_CONTEXT_BASE_SV + PLIC_CONTEXT_CLAIM)
+
+typedef struct plic_state {
+  rman_t rm;                 /* irq resource manager */
+  resource_t *mem;           /* PLIC memory resource */
+  resource_t *irq;           /* PLIC irq resource */
+  intr_event_t **intr_event; /* interrupt events */
+  unsigned ndev;             /* number of sources */
+} plic_state_t;
+
+#define in4(addr) bus_read_4(plic->mem, (addr))
+#define out4(addr, val) bus_write_4(plic->mem, (addr), (val))
+
+static void plic_intr_disable(intr_event_t *ie) {
+  plic_state_t *plic = ie->ie_source;
+  unsigned irq = ie->ie_irq;
+
+  uint32_t en = in4(PLIC_ENABLE_SV(irq));
+  en &= ~(1 << (irq % 32));
+  out4(PLIC_ENABLE_SV(irq), en);
+}
+
+static void plic_intr_enable(intr_event_t *ie) {
+  plic_state_t *plic = ie->ie_source;
+  unsigned irq = ie->ie_irq;
+
+  uint32_t en = in4(PLIC_ENABLE_SV(irq));
+  en |= 1 << (irq % 32);
+  out4(PLIC_ENABLE_SV(irq), en);
+}
+
+static const char *plic_intr_name(unsigned irq) {
+  char buf[32];
+  snprintf(buf, sizeof(buf), "PLIC source %u", irq);
+  return kstrndup(M_STR, buf, sizeof(buf));
+}
+
+static resource_t *plic_alloc_intr(device_t *pic, device_t *dev, int rid,
+                                   unsigned irq, rman_flags_t flags) {
+  plic_state_t *plic = pic->state;
+  rman_t *rman = &plic->rm;
+
+  return rman_reserve_resource(rman, RT_IRQ, rid, irq, irq, 1, 0, flags);
+}
+
+static void plic_release_intr(device_t *pic, device_t *dev, resource_t *r) {
+  resource_release(r);
+}
+
+static void plic_setup_intr(device_t *pic, device_t *dev, resource_t *r,
+                            ih_filter_t *filter, ih_service_t *service,
+                            void *arg, const char *name) {
+  plic_state_t *plic = pic->state;
+  unsigned irq = resource_start(r);
+  assert(irq && irq < plic->ndev);
+
+  if (!plic->intr_event[irq])
+    plic->intr_event[irq] = intr_event_create(
+      plic, irq, plic_intr_disable, plic_intr_enable, plic_intr_name(irq));
+
+  r->r_handler =
+    intr_event_add_handler(plic->intr_event[irq], filter, service, arg, name);
+}
+
+static void plic_teardown_intr(device_t *pic, device_t *dev, resource_t *r) {
+  intr_event_remove_handler(r->r_handler);
+}
+
+static int plic_map_intr(device_t *pic, device_t *dev, phandle_t *intr,
+                         int icells) {
+  plic_state_t *plic = pic->state;
+
+  if (icells != 1)
+    return -1;
+
+  unsigned irq = *intr;
+  if (!irq || irq > plic->ndev)
+    return -1;
+
+  return irq;
+}
+
+static intr_filter_t plic_intr_handler(void *arg) {
+  plic_state_t *plic = arg;
+
+  /* Claim any pending interrupt. */
+  uint32_t irq = in4(PLIC_CLAIM_SV);
+
+  if (irq) {
```

Comment:
I prefer if-clause to filter out weird cases leaving the main task of function to be implemented at the very bottom.

Reply:
PLIC spec: "An interrupt ID of 0 is reserved to mean “no interrupt”."
I changed the control flow and inserted a panic on this faulty condition.